### PR TITLE
make raw socket usable and minor optimization.

### PR DESCRIPTION
**Pull request by urbanishimwe**
**ListenPacket** can be used to listen to traffics from networks that are not **multicast**. it takes less time and fewer resources to spin up **raw socket** more than **pcap**.
The implementations applied is to create an active routine that continuously read packet as they come.
For requests of the large body, one may increase the buffer size. the default/minimum buffer size is 64kb(this is because of the MTU of most non-multicast IP is 64kb).

These changes were tested using **postman**

### need help
- it doesn't work with `curl, browser....`

### CODE REVIEW

File: capture/listener.go [hunk 1]
```diff
@@ -83,7 +75,7 @@ type Listener struct {
 	pcapHandles []*pcap.Handle
 
 	quit    chan bool
-	readyCh chan bool
+	readyCh bool
```
Comment (slimus) [inline capture/listener.go:78]:
```suggestion
	ready bool
```

File: capture/listener.go [hunk 2]
```diff
@@ -526,7 +520,7 @@ func (t *Listener) readPcap() {
 					continue
 				}
 
-				dataOffset := (data[12] & 0xF0) >> 4
+				dataOffset := data[12] >> 4
```
Comment (slimus) [inline capture/listener.go:523]: A lot of magic  :) 
Can you explain what did\does it mean?

File: capture/listener.go [hunk 3]
```diff
@@ -586,7 +580,9 @@ func (t *Listener) readPcap() {
 	}
 
 	wg.Wait()
-	t.readyCh <- true
+	t.Lock()
+	t.readyCh = true
```
Comment (slimus) [inline capture/listener.go:584]:
```suggestion
	t.ready = true
```

File: capture/listener.go [hunk 4]
```diff
@@ -600,7 +596,9 @@ func (t *Listener) readPcapFile() {
 			}
 		}
 
-		t.readyCh <- true
+		t.Lock()
+		t.readyCh = true
```
Comment (slimus) [inline capture/listener.go:600]:
```suggestion
		t.ready = true
```

File: capture/listener.go [hunk 5]
```diff
@@ -640,7 +638,7 @@ func (t *Listener) readPcapFile() {
 				continue
 			}
 
-			dataOffset := (data[12] & 0xF0) >> 4
+			dataOffset := data[12] >> 4
```
Comment (slimus) [inline capture/listener.go:641]: Copy\paste from https://github.com/buger/goreplay/pull/768/files#diff-cb5774656d51d7a2d177f84ee666c512R523.
Is it possible to move to separate func?

File: capture/listener.go [hunk 6]
```diff
@@ -657,32 +655,45 @@ func (t *Listener) readPcapFile() {
...
-
+	type RSPacket struct {
+		buf  []byte
+		addr net.Addr
+		err  error
+		n    int
+	}
+	var bufChan = make(chan *RSPacket, 1000)
+	t.Lock()
+	t.readyCh = true
```
Comment (slimus) [inline capture/listener.go:670]:
```suggestion
	t.ready = true
```

File: capture/listener.go [hunk 7]
```diff
@@ -657,32 +655,45 @@ func (t *Listener) readPcapFile() {
...
+	t.Unlock()
+	go func() {
+		buffer := 64 * 1024
```
Comment (slimus) [inline capture/listener.go:673]: Create const like `defaultBufferSize = 64 * 1024` and use it here.

File: capture/listener.go [hunk 8]
```diff
@@ -701,16 +712,13 @@ func (t *Listener) isValidPacket(buf []byte) bool {
 	// http://en.wikipedia.org/wiki/Transmission_Control_Protocol
 	destPort := binary.BigEndian.Uint16(buf[2:4])
 	srcPort := binary.BigEndian.Uint16(buf[0:2])
-
 	// Because RAW_SOCKET can't be bound to port, we have to control it by ourself
 	if destPort == t.port || (t.trackResponse && srcPort == t.port) {
 		// Get the 'data offset' (size of the TCP header in 32-bit words)
-		dataOffset := (buf[12] & 0xF0) >> 4
-
+		dataOffset := buf[12] >> 4
```
Comment (slimus) [inline capture/listener.go:718]: and again copy\paste...

File: capture/listener_test.go [hunk 9]
```diff
@@ -12,7 +12,7 @@ import (
 func TestRawListenerInput(t *testing.T) {
 	var req, resp *TCPMessage
 
-	listener := NewListener("", "0", EnginePcap, true, 10*time.Millisecond, "", "", 0, false, false)
+	listener := NewListener("", "0", EnginePcapFile, true, 10*time.Millisecond, "", "", 0, false, false)
```
Comment (slimus) [inline capture/listener_test.go:15]: Is it correct? Why we don't test EnginePcap? May be we should add the new test?

File: capture/listener_test.go [hunk 10]
```diff
@@ -439,30 +431,34 @@ func testChunkedSequence(t *testing.T, listener *Listener, packets ...*TCPPacket
...
-
-	copy(next, append(list[:k], list[k+1:]...))
-
-	return append(first, permutation(n/len(list), next)...)
 }
 
 // Response comes before Request
 func TestRawListenerChunkedWrongOrder(t *testing.T) {
-	listener := NewListener("", "0", EnginePcap, true, 10*time.Millisecond, "", "", 0, false, false)
+	listener := NewListener("", "0", EnginePcap, true, 10*time.Second, "", "", 0, false, false)
```
Comment (slimus) [inline capture/listener_test.go:461]: Why did you change expire to 10 sec?

Comment (urbanishimwe) [reply capture/listener_test.go:461]: The parsing of packets was taking longer than `1o milliseconds` but it won't spend `10s` its to give it enough time and the logic of the tests will not be affected

File: capture/listener.go (hunk 2 above)
Comment (urbanishimwe) [reply capture/listener.go:522]: in TCP Header the data offset is identified at `the last 4 bits of 12th byte` thats why we've done this shifting

File: capture/listener.go (hunk 5 above)
Comment (urbanishimwe) [reply capture/listener.go:640]: moving it to another function will cause an un-necessary increase in CPU cycles which will decrease performance!

Comment (slimus) [reply capture/listener.go:640]: did you try it? so now it ok, but can you please create some example to show increase CPU and decrease performance?
